service/service.go

```diff
@@ -170,59 +173,96 @@ func NewService(cfg *Config, datadir string) (*Service, error) {
 		privKey:         privateKey,
 		PubKey:          privateKey.Public().(ed25519.PublicKey),
 	}
+
 	log.Info("Service public key: %x", s.PubKey)
 
 	return s, nil
 }
 
-func (s *Service) loop(ctx context.Context) {
-	var executingRounds errgroup.Group
-	defer executingRounds.Wait()
+type roundResult struct {
+	round *round
+	err   error
+}
+
+func (s *Service) loop(ctx context.Context, roundsToResume []*round, cmds <-chan Command) {
+	var eg errgroup.Group
+	defer eg.Wait()
+
+	logger := log.AppLog.WithName("worker")
+	ctx = logging.NewContext(ctx, logger)
+
+	roundResults := make(chan roundResult, 1)
+
+	// Resume recovered rounds
+	for _, round := range roundsToResume {
+		round := round
+		s.executingRounds[round.ID] = struct{}{}
+		end := s.roundEndTime(round)
+		eg.Go(func() error {
+			err := round.recoverExecution(ctx, round.stateCache.Execution, end)
+			roundResults <- roundResult{round: round, err: err}
+			return nil
+		})
+	}
+
+	timer := s.scheduleRound(ctx, s.openRound)
 
 	for {
-		s.openRoundMutex.RLock()
-		epoch := s.openRound.Epoch()
-		s.openRoundMutex.RUnlock()
-
-		start := s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift)
-		waitTime := time.Until(start)
-		timer := time.After(waitTime)
-		if waitTime > 0 {
-			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), waitTime)
-		}
 		select {
+		case cmd := <-cmds:
+			cmd(s)
+
+		case result := <-roundResults:
+			if result.err == nil {
+				broadcaster := s.broadcaster
+				go broadcastProof(s, result.round, result.round.execution, broadcaster)
+			}
+			delete(s.executingRounds, result.round.ID)
+
 		case <-timer:
+			round := s.openRound
+			s.openRound = s.newRound(ctx, round.Epoch()+1)
+			s.executingRounds[round.ID] = struct{}{}
+
+			end := s.roundEndTime(round)
+			minMemoryLayer := s.minMemoryLayer
+			eg.Go(func() error {
+				roundResults <- executeRound(ctx, round, end, minMemoryLayer)
+				return nil
+			})
+
+			// schedule the next round
+			timer = s.scheduleRound(ctx, s.openRound)
+
 		case <-ctx.Done():
-			log.Info("service shutting down")
-			s.openRoundMutex.Lock()
-			s.openRound = nil
-			s.openRoundMutex.Unlock()
+			logger.Info("service shutting down")
 			return
 		}
+	}
+}
 
-		s.openRoundMutex.Lock()
-		prevRound := s.openRound
-		s.newRound(ctx, prevRound.Epoch()+1)
-		s.openRoundMutex.Unlock()
+func (s *Service) roundStartTime(round *round) time.Time {
+	return s.genesis.Add(s.cfg.PhaseShift).Add(s.cfg.EpochDuration * time.Duration(round.Epoch()))
+}
 
-		executingRounds.Go(func() error {
-			round := prevRound
-			if err := s.executeRound(ctx, round); err != nil {
-				s.asyncError(fmt.Errorf("round %v execution error: %v", round.ID, err))
-				return nil
-			}
-			broadcastProof(s, round, round.execution, s.getBroadcaster())
-			return nil
-		})
+func (s *Service) roundEndTime(round *round) time.Time {
+	return s.roundStartTime(round).Add(s.cfg.EpochDuration).Add(-s.cfg.CycleGap)
+}
```

Comment:
I feel like these two should be methods of `round`: `round::Start()` and `round::End()`.

Reply:
A `round` isn't aware of genesis, epoch duration, phase shift etc. so it couldn't calculate when to start and when to end.

Reply:
That's an interesting idea. @noam also pointed out recently that rounds/poet shouldn't be aware of epochs. Anyway - `round` refactoring is out of the scope of this PR. I will be refactoring `round` separately (I want to de-async its teardown logic and reduce the number of channels there).

Reply:
Could you create an issue for that or add it as requirement to an existing issue?

Reply:
Done: https://github.com/spacemeshos/poet/issues/181